README.md
# gui

A Clojure library that serves as a meta-library/wrapper around other
libraries and convenience packages, to ensure the widest range of GUI
and utility support across Java versions and operating systems.

In general, I plan to ensure compatability with all JRE from Java 8 to
Java 12+ for the packages that are installed as dependencies (this
obviously means versions of some packages will be out of date -
usually that's fine, particularly if we weren't using some feature of
them anyways).

The main bullet-points I want to hit with this are common things that
many GUI interfaces will tend to require, but may not cleanly fall
into the Clojure built in facilities:

- GUI rendering
- GUI Theming/Styling
- Networking + Serialization
- Logging
- Data Persistence (config file based)

![light](https://github.com/ahungry/gui/blob/master/light-gui.png)
![dark](https://github.com/ahungry/gui/blob/master/dark-gui.png)

# Chosen packages/versions

These have been tested and known to work, on a range of JRE
environments, from Windows 7 (32 bit) + Oracle JRE 8 (last 32 Oracle
one), to Win7 + Zulu OpenJDK JRE 8 to 12, as well as Arch Linux and
Ubuntu OpenJDKs 8 to 12.

## Main layout/GUI interface: Swing + Clojure Seesaw

See notes for some details into why this sort of won by default.

- Clojure Seesaw: https://github.com/daveray/seesaw
- Styles + enhancements to swing: https://github.com/kirill-grouchnikov/radiance

Some of the packages (radiance) had to be installed at their 1.x
versions, as they are the last ones that had maven jars compiled to
work with Java API version 52 (aka, Java 8).

Radiance has a great theming tool, that you can see in the demo when
you boot this app up.

# Notes

Considered but didn't work out:

- JavaFX is a slick library, but support for it across Java versions
  is very hard to align.  Oracle Java 8 ships with it built in,
  OpenJDK 8 ships it as a separate package on GNU/Linux
  distributions.  Later versions of Java (9 and 10) don't seem to have
  it in a clear location.  Java 11 has it as a separate Maven
  installable library.  It might be a good choice for a cross-version
  / OS jar, *if* you get to dictate what Java version your potential
  userbase must use (or if it was for an internal app where you
  provision the appropriate software on the target machines).
- cljfx (https://github.com/cljfx/cljfx) is a great library, but has
  an intentional design choice to only support the latest underlying
  systems.  As such, it was complicated to get working on Arch Linux
  as well as Windows.  The fact it is closely tied to JavaFX also
  makes it difficult to use.
- fn-fx (https://github.com/fn-fx/fn-fx) has similar problems to
  cljfx - very hard to use across a wide-berth of targets (see the
  usage matrix on their github readme for a testament to that - way
  too many red Xs)

## Notes - Setting up profile to test different java version

On ArchLinux you can use `sudo archlinux-java set java-8-openjdk/jre`
or `sudo archlinux-java set java-12-openjdk` (Ubuntu has an update
alternatives approach).

On Windows, you can adjust in your ~/.lein/profile.clj by adding
something like this:

```clojure
{:user
  {
    :java-cmd "C:\\Program\ Files\\Java\\jdk1.8.0_202\\bin\\java.exe "
    ;; :java-cmd "C:\\Users\\IEUser\\scoop\\apps\\zulu12\\current\\bin\\java.exe"
  }
}
```

# TODO

- Integrate xdg-rc package (add windows support there) to allow the
  'default' gui view / package to easily incorporate the radiant theme
  switcher and save user setting for this.
- Figure out keybind to quickly cycle through active tabs (alt+N to go
  to that tab, ctrl + tab / ctrl + shift + tab to iterate across)
- Figure out how to add X to tab to close it out

# License

Copyright © 2019 Matthew Carter <m@ahungry.com>

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-09-15
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-09-15
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/gui/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/gui/compare/0.1.0...0.1.1


test/ahungry/gui_test.clj
(ns ahungry.gui-test
  (:require
   [clojure.test :refer :all]
   [clojure.tools.logging :as log]
   [ahungry.net :as net]
   [ahungry.gui :refer :all]
   [xdg-rc.core :as xdg-rc]
   ))

(deftest log-test
  (testing "The log call works and returns nil."
    (is (= nil (log/debug "Greetings from the test suite.
The file that controls log output is in:

  ./src/logback.xml

You can edit the verbosity level in there.")))))

(deftest net-test
  (testing "That we can pull some remote data with our chosen http libs."
    (let [res (net/get-json "https://httpbin.org/ip" {})]
      (println "Your IP was: ")
      (prn res)
      (is (= java.lang.String (type (:origin res)))))))

(deftest config-test
  (testing "That we can use this tool for this."
    (let [res (xdg-rc/get-xdg-config-home)]
      (println "Your home directory was...:")
      (prn res)
      (is (= java.lang.String (type res))))))


test/ahungry/keys_test.clj
(ns ahungry.keys-test
  (:require
   [clojure.test :refer :all]
   [ahungry.keys :refer :all]
   ))

(deftest code->key-test
  (testing "We can convert a java.awt.event.KeyEvent to a readable key kw."
    (is (= :shift (code->key 16)))
    (is (= :ctrl (code->key 17)))
    (is (= :meta (code->key 18)))
    (is (= :super (code->key 524)))
    (is (= \A (code->key 65)))))

(deftest modkey?-test
  (testing "Modkeys should show up as such."
    (is (not (modkey? :x)))
    (is (modkey? :shift))))

(deftest modchar->modkey-test
  (testing "We can convert strings to the proper kw."
    (is (= :ctrl (modchar->modkey "C")))
    (is (= :meta (modchar->modkey "M")))
    (is (= :super (modchar->modkey "S")))
    (is (= "x" (modchar->modkey "x")))))

(deftest keystring->keybind-test
  (testing "We can do the conversions."
    (is (= #{:ctrl :meta "f"} (keystring->keybind "C-M-f")))))

(deftest active-modkeys-test
  (testing "We can extract active keys."
    (is (= #{:ctrl :meta} (active-modkeys {:ctrl true :meta true :super nil})))))

(deftest is-keyequal?-test
  (testing "The equality comparisons hold up."
    (is (= true (is-keyequal? "m" \m {:ctrl false :meta false})))
    (is (= false (is-keyequal? "C-m" \m {:ctrl true :meta true})))
    (is (= true (is-keyequal? "C-M-b" \b {:ctrl true :meta true})))
    (is (= true (is-keyequal? "M-m" \m {:ctrl false :meta true})))))

(deftest dispatch-get-fns-test
  (testing "That functions or values we expect to return do so."
    (let [keymap {"a" 42
                  "C-b" 52
                  "C-M-x" 90}]
      (is (= 42 (first (dispatch-get-fns keymap \a {:ctrl nil}))))
      (is (= 52 (first (dispatch-get-fns keymap \b {:ctrl true}))))
      (is (= 90 (first (dispatch-get-fns keymap \x {:ctrl true :meta true}))))
      (is (= '() (dispatch-get-fns keymap \z {})))
      )
    ))


project.clj
(defproject ahungry/gui "0.1.0-SNAPSHOT"
  :description "GUI packages chosen to support a wide range of JRE environments."
  :url "https://github.com/ahungry/gui"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.0"]
                 [seesaw "1.5.0"]

                 ;; Latest versions, but only work with java 9+
                 ;; [org.pushing-pixels/radiance-substance "2.5.1"]
                 [org.pushing-pixels/radiance-substance "1.0.2"]

                 ;; Logging related
                 [ch.qos.logback/logback-classic "1.2.3"]
                 [org.clojure/tools.logging "0.5.0"]

                 ;; Network related
                 [clj-http "3.10.0"]
                 [cheshire "5.9.0"]

                 ;; Config or persistence related
                 [ahungry/xdg-rc "0.0.4"]

                 ]
  :main ahungry.gui
  :repl-options {:init-ns ahungry.gui}
  :java-source-paths ["java-src"]
  :aot [ahungry.gui]
  :target-path "target/%s"
  :profiles
  {
   ;; Required to use clojure spec check facilities.
   :dev {:dependencies [[org.clojure/test.check "0.9.0"]]}
   :uberjar {:aot :all}
   }
  :jvm-opts ["-Dfile.encoding=UTF8"
             "-Dswing.aatext=true"
             "-Dawt.useSystemAAFontSettings=lcd"
             ;; "-Dswing.defaultlaf=org.pushingpixels.substance.api.skin.DustSkin"
             ]
  :pom-addition [:properties
                 ["maven.compiler.source" "1.8"]
                 ["maven.compiler.target" "1.8"]])


src/ahungry/gui.clj
(ns ahungry.gui
  (:require
   [seesaw.core :as ss]
   [seesaw.keystroke :as sk]
   [seesaw.graphics :as ssg]
   [seesaw.color :as ssc]
   [seesaw.action :as ssa]
   [seesaw.config :as ssconfig]
   [clojure.tools.logging :as log]
   [ahungry.net :as net]
   [ahungry.keys :as keys]
   )
  (:use
   [seesaw.core]
   [seesaw.graphics]
   [seesaw.color]
   )
  (:import org.pushingpixels.substance.api.SubstanceCortex$GlobalScope)
  (:gen-class))

;; TODO: Programatically set log level based on boot up or something.
;; :trace, :debug, :info, :warn, :error, :fatal
(log/debug "Boot")

;; Tries to match styles to native host some
(ss/native!)

(set! *warn-on-reflection* false)

;; (def normal-font "ARIAL-12-PLAIN")
;; (def normal-font "ARIAL-20-PLAIN")
(def normal-font "ARIAL-PLAIN-20")
(def title-font "ARIAL-14-BOLD")

(defn laf-selector []
  (ss/horizontal-panel
   :items ["Substance skin: "
           (ss/combobox
            :model    (vals (SubstanceCortex$GlobalScope/getAllSkins))
            :renderer (fn [this {:keys [value]}]
                        (ss/text! this (.getClassName value)))
            :listen   [:selection (fn [e]
                                        ; Invoke later because CB doens't like changing L&F while
                                        ; it's doing stuff.
                                    (ss/invoke-later
                                     (-> e
                                         ss/selection
                                         .getClassName
                                         SubstanceCortex$GlobalScope/setSkin)))])]))

(def notes " This example shows the available Substance skins. Substance
  is a set of improved look and feels for Swing. To use it in a project,
  you'll need to add a dep to your Leiningen project:

        [com.github.insubstantial/substance \"7.1\"]

  In this example, the full class name of the current skin is shown the
  in the combobox above. For your own apps you could either use a
  selector like this example, or, more likely, set a default initial
  skin in one of the following ways:

    Start your VM with -Dswing.defaultlaf=<class-name>

    Call (javax.swing.UIManager/setLookAndFeel \"<class-name>\")
    do this *after* (seesaw.core/native!) since that sets the L&F.

  See http://insubstantial.github.com/insubstantial/substance/docs/getting-started.html
  for more info. There you'll also find much more info about the
  skins along with much less crappy looking demos.")

;; BEGIN paint sample from canvas.clj in seesaw repo
(defn paint1 [c g]
  (let [w (.getWidth c)
        h (.getHeight c)]
    (doto g
      (ssg/draw (ssg/polygon [0 h] [(/ w 4) 0] [(/ w 2) (/ h 2)] [w (/ h 2)] [0 h])
                (ssg/style :foreground java.awt.Color/BLACK
                           :background (ssc/color 128 128 128 128)
                           :stroke     (ssg/stroke :width 4)))
      (.setColor (ssc/color 224 224 0 128))
      (.fillRect 0 0 (/ w 2) (/ h 2))
      (.setColor (ssc/color 0 224 224 128))
      (.fillRect 0 (/ h 2) (/ w 2) (/ h 2))
      (.setColor (ssc/color 224 0 224 128))
      (.fillRect (/ w 2) 0 (/ w 2) (/ h 2))
      (.setColor (ssc/color 224 0 0 128))
      (.fillRect (/ w 2) (/ h 2) (/ w 2) (/ h 2))
      (.setColor (ssc/color 0 0 0))
      (.drawString "Hello. This is a canvas example" 20 20))))

(defn get-image [filename]
  (javax.imageio.ImageIO/read (clojure.java.io/file filename)))

(defn xy-looper
  "Loop across a grid of W width and H height in steps of X and Y, applying f(x y) each time."
  [w h x y f]
  (loop [ix 0]
    (loop [iy 0]
      (f (* x ix) (* y iy))
      (when (< (* y iy) h) (recur (inc iy))))
    (when (< (* x ix) w) (recur (inc ix)))))

(defn paint-image [c graphics]
  (let [w (.getWidth c)
        h (.getHeight c)
        x 256    ; This should be dynamic from image, not hardcoded...
        y 256    ; This should be dynamic from image, not hardcoded...
        img (get-image "main-bg.jpg")]
    (xy-looper w h x y (fn [x y] (.drawImage graphics img nil x y)))))

(def text-style (ssg/style :foreground (ssc/color 0 0 0)
                           :font "ARIAL-BOLD-24"))

;; These calls have to be unqualified because the macro is just converting
;; the symbols into method invocations of the thing being painted.
(def star
  (ssg/path []
            (move-to 0 20) (line-to 5 5)
            (line-to 20 0) (line-to 5 -5)
            (line-to 0 -20) (line-to -5 -5)
            (line-to -20 0) (line-to -5 5)
            (line-to 0 20)))

(def grow (atom {:x 0 :y 0 :f inc}))

(defn get-grow []
  (let [{:keys [x y f]} @grow]
    (when (> x 300)
      (swap! grow assoc-in [:f] dec))
    (when (< x 1)
      (swap! grow assoc-in [:f] inc))
    (swap! grow conj {:x (f x) :y (f y)})
    @grow))

(defn paint2 [c g]
  (let [grow (get-grow)
        ;; w (.getWidth c)  w2 (/ w 2)
        ;; h (.getHeight c) h2 (/ h 2)
        w2 (:x grow)
        h2 (:y grow)
        ]
    (ssg/draw g
              (ssg/ellipse 0  0  w2 h2) (ssg/style :background (ssc/color 224 224 0 128))
              (ssg/ellipse 0  h2 w2 h2) (ssg/style :background (ssc/color 0 224 224 128))
              (ssg/ellipse w2 0  w2 h2) (ssg/style :background (ssc/color 224 0 224 128))
              (ssg/ellipse w2 h2 w2 h2) (ssg/style :background (ssc/color 224 0 0 128)))
    (ssg/push g
              (ssg/rotate g 20)
              (ssg/draw g (ssg/string-shape 20 20  "Hello. This is a canvas example") text-style))
    (ssg/push g
              (ssg/translate g w2 h2)
              (ssg/draw g star (ssg/style :foreground java.awt.Color/BLACK :background java.awt.Color/YELLOW)))))

                                        ; Create an action that swaps the paint handler for the canvas.
                                        ; Note that we can use (config!) to set the :paint handler just like
                                        ; properties on other widgets.
(defn switch-paint-action [n paint]
  (ssa/action :name n
              :handler #(-> (ss/to-frame %)
                            (ss/select [:#canvas])
                            (ssconfig/config! :paint paint))))
;; END paint sample from canvas.clj in seesaw repo

(defn make-canvas-panel []
  (ss/canvas :id :canvas1
             :background "#BBBBDD"
             :paint paint1))

(defn make-canvas-panel2 []
  (ss/canvas :id :canvas2
             :background "#BBBBDD"
             :paint paint2))

;; We can do a dynamic or REPL based flow with config!
;; (def x (make-canvas-panel2))
;; (show x)
;; (ss/config! x :paint 1)
;; (ss/config! x :paint 2)
;; (while true (do (Thread/sleep 10) (ss/config! x :paint paint2)))

(defn make-tabbed-pane []
  (ss/tabbed-panel :tabs [{:title "One" :content (make-canvas-panel)}
                          {:title "Two" :content (make-canvas-panel2)}]))

(defn a-test [e]
  (prn e)
  (ss/alert "Hello"))

;; Set up a root we can ref later as an atom.
(def *root (atom nil))

;; NOTE: hmm
;; Hmm, it may be easiest to do all "top level" bindings in the menu area.
;; Otherwise, it seems like certain elements can eat all the custom keybinds/listeners.
(defn make-menu []
  (let [a-test (ssa/action :handler a-test :name "Test" :tip "Pop up an alert" :key "menu A")
        tab1 (ssa/action :handler (fn [_e] (ss/selection! @*root 1))
                         :name "Select Tab 1"
                         :tip "Jump to tab 1."
                         :key "menu 1")
        tab2 (ssa/action :handler (fn [_e] (ss/selection! @*root 2))
                         :name "Select Tab 1"
                         :tip "Jump to tab 1."
                         :key "menu 2")
        ]
    (ss/menubar
     :items [(ss/menu :text "File" :items [a-test])
             (ss/menu :text "Tabs" :items [tab1 tab2])])))

(defn x-action-handler [e]
  (prn e))

(def x-action (action :name "XA" :key "alt 2" :handler x-action-handler))

(defn add-behaviors [root]
  (ss/config! (select root [:.xa]) :action x-action)
  root)

(defn make-switchable-canvas []
  (ss/vertical-panel
   :class :xa
   :items
   [(ss/canvas :id :canvas :background "#BBBBDD" :paint paint1)
    (ss/horizontal-panel :items ["Switch canvas paint function: "
                                 (switch-paint-action "None" nil)
                                 (switch-paint-action "Rectangles" paint1)
                                 (switch-paint-action "Ovals" paint2)])]))

(defn make-laf-stuff []
  (ss/border-panel
   ;; :class :xa
   :hgap 5 :vgap 5 :border 5
   :center (ss/vertical-panel
            :paint paint-image
            :items [
                    :separator
                    (laf-selector)
                    (ss/text :multi-line? true
                             :text notes
                             :border 5
                             :font normal-font
                             :opaque? false
                             :foreground "#ffffff"
                             ;; Yes! we can paint custom.
                             ;; Would it work with image?
                             ;; Well, almost - the painted image covers up the text...
                             ;; :paint paint-image
                             )
                    :separator
                    (ss/label :text "A Label")
                    (ss/button :text "A Button")
                    (ss/checkbox :text "A checkbox")
                    (ss/combobox :model ["A combobox" "more" "items"])
                    (ss/horizontal-panel
                     :border "Some radio buttons"
                     :items (map (partial ss/radio :text)
                                 ["First" "Second" "Third"]))
                    (ss/scrollable (ss/listbox :model (range 100)))])))

(defn make-main []
  (ss/tabbed-panel
   :tabs
   [
    {:title "Look and Feel" :content (make-laf-stuff)}
    {:title "Switchable Canvas" :content (make-switchable-canvas)}
    {:title "Paint1" :content (make-canvas-panel)}
    {:title "Paint2"
     ;; :icon (slurp "close-icon.png")
     :content (make-canvas-panel2)}
    ]))

;; (ss/listen x :key-pressed (fn [e] (prn e)))
;; Programatically interact with tabs via:
;; (ss/selection x) ; Get the selection
;; (ss/selection! x 0) ; Set the selection

(defn xset-listeners!
  "Uses Emacs style key binding maps to bind user button presses to
  functions to invoke on key pressed events.  We would probably want
  to update this to handle sending args to the events in some way also."
  [x]
  (ss/listen x :key-released keys/handle-key-released)
  (ss/listen x :key-pressed (keys/handle-key-pressed keys/global-keymap)))

(defn dispatch-event [^java.awt.event.KeyEvent e]
  (try
    (let [id (.getID e)]
      (cond
        (= id java.awt.event.KeyEvent/KEY_PRESSED)
        ((keys/handle-key-pressed keys/global-keymap) e)

        (= id java.awt.event.KeyEvent/KEY_RELEASED)
        (keys/handle-key-released e)

        :else
        nil
        ;; (prn "Some unmapped key event...")
        ))
    (catch Exception e (prn e))))

(defn set-listeners!
  "Global listener handling irregardless of what is focused."
  []
  (let [manager (java.awt.KeyboardFocusManager/getCurrentKeyboardFocusManager)
        dispatcher (reify java.awt.KeyEventDispatcher
                     (dispatchKeyEvent [this e]
                       (dispatch-event e)
                       false))]
    (doto manager
      (.addKeyEventDispatcher dispatcher))))

(defonce listeners (set-listeners!))

(defn draw-a-red-x
  "Draw a red X on a widget with the given graphics context"
  [c g]
  (let [w          (ss/width c)
        h          (ss/height c)
        line-style (ssg/style :foreground "#FF0000" :stroke 3 :cap :round)
        d 5]
    (ssg/draw g
              (ssg/line d d (- w d) (- h d)) line-style
              (ssg/line (- w d) d d (- h d)) line-style)))

(defn make-red-x []
  (ss/flow-panel
   :border 5
   :items [
           (ss/label  :text "I'm a good label!" :font "ARIAL-BOLD-40" :foreground "#00AA00")
           (ss/label  :text "I'm a bad label!"  :font "ARIAL-BOLD-40" :paint draw-a-red-x)
           (ss/button :text "I'm a bad button!"  :font "ARIAL-BOLD-40" :paint draw-a-red-x)]))

;; look in seesaw text_editor sample for a menu bar and probably key bindings
(defn show
  "REPL friendly way to pop up what we're working on."
  [f]
  (keys/init!)
  (ss/invoke-later
   (->
    (ss/frame
     :minimum-size [640 :by 480]
     :menubar (make-menu)
     :title "Widget"
     :content f)
    ;; add-behaviors
    ss/pack!
    ss/show!)
   ;; (set-listeners! f)
   ))

;; Just debug WIP stuff
(def x (make-main))
(reset! *root x)
;; (set-listeners! x)
;; Ensure we don't eat up the tab key presses
(.setFocusTraversalKeysEnabled x false)

(defn set-root! [x] (reset! *root x) x)

(defn -main [& args]
  (ss/invoke-later
   (->
    (ss/frame
     :title "Seesaw Substance/Insubstantial Example"
     :minimum-size [640 :by 480]
     :menubar (make-menu)
     :on-close :exit
     :content
     (-> (make-main) set-root!)
     )
    ss/pack!
    ss/show!)
   ;; Calling this, or setting it via REPL causes some issues...
   (SubstanceCortex$GlobalScope/setSkin "org.pushingpixels.substance.api.skin.NebulaSkin")
   ))

(log/debug "Fin")


src/ahungry/keys.clj
;; Facilities for Emacs-like key handling.

(ns ahungry.keys
  (:require
   [clojure.tools.logging :as log]
   [ahungry.util :as util]
   ))

;; Maps from Swing
;; key-pressed key-released key-typed
;; 16 - shift
;; 18 - alt
;; 17 - ctrl
;; 27 - esc
;; 524 - super left
;; 38 - ARROW_UP
;; 37 - ARROW_LEFT (by default moves through tab panes)
;; 39 - ARROW_RIGHT
;; 40 - ARROW_DOWN

(def *modkeys (atom {:ctrl nil
                     :meta nil
                     :shift nil
                     :super nil}))

(defn init!
  "Resets atom state."
  []
  (reset! *modkeys {:ctrl nil :meta nil :shift nil :super nil}))

(defn code->key
  "Receive a java.awt.event.KeyEvent key code N, turn it into the readable key."
  [n]
  (case n
    16 :shift
    17 :ctrl
    18 :meta
    524 :super
    9 :tab
    32 :space
    (char n)))

(defn modkey? [kw] (contains? @*modkeys kw))
(defn e->char [e] (.getKeyChar e))
(defn e->code [e]
  ;; (log/debug "Translating event to code char: " (.getKeyCode e))
  (.getKeyCode e))
(def e->key (comp code->key e->code))

(defn modchar->modkey [s]
  (case s
    "C" :ctrl
    "M" :meta
    "S" :super
    s))

(defn keystring->keybind
  "Turns something like C-M-f into #{:ctrl :meta \"f\"}."
  [s]
  (->> (clojure.string/split s #"-")
       (map modchar->modkey)
       set))

(defn active-modkeys
  "Just pull out/filter the active modkeys and return a set of them."
  [m]
  (->> (into [] m)
       (filter (fn [[_ v]] (= true v)))
       (map first)
       set))

(defn case-insensitive-eq
  "Compare strings S1 and S2 in a case insensitive manner."
  [s1 s2]
  (= (.toLowerCase s1) (.toLowerCase s2)))

;; TODO: Probably split keys by spaces to allow nested keys.
(defn is-keyequal?
  "Given an Emacs like key binding S (such as C-f (ctrl + f) or M-1 (meta
  + 1)), see if it is equal to the code C, while considering the
  associated modkeys states M."
  [s c m]
  (let [keybind (keystring->keybind s)
        keybind-base (first (filter (complement modkey?) keybind))
        keybind-mods (filter modkey? keybind)
        active-modkeys (active-modkeys m)
        all-mods-match? (util/contains-all? keybind-mods active-modkeys)]
    (and
     (= (count active-modkeys) (count keybind-mods))
     all-mods-match?
     (case-insensitive-eq keybind-base (str c)))))

(defn handle-key-released
  "Unset key handling facilities."
  [e]
  (let [key (e->key e)]
    (when (modkey? key)
      ;; (log/debug "Modkey is released, unsetting it.")
      (swap! *modkeys assoc-in [key] false))))

(defn foo [] (log/info "Foo was called."))
(defn bar [] (log/info "Bar was called."))

(def global-keymap
  {"a" #'foo
   "C-b" #'bar})

(defn filter-by-is-keyequal?
  "Expects col to be a col of tuples, where the first is the binding."
  [c state]
  (fn [col]
    (let [keybind (first col)]
      (is-keyequal? keybind c state))))

(defn dispatch-get-fns
  "Receives a map M of keybinds, a char C, and works against active STATE to find
  which keybind is appropriate to return a function for."
  [m c state]
  (->> (into [] m)
       (filter (filter-by-is-keyequal? c state))
       (map second)
       doall))

(defn dispatch
  "Receives a map M of keybinds, a char C, and works against active STATE to find
  which keybind is appropriate to return a function for and calls them."
  [m c state]
  (let [fs (dispatch-get-fns m c state)]
    ;; (log/debug "Dispatch came up with this list of functions to run: " fs)
    fs))

(defn handle-key-pressed
  "Set key handling facilities.  Receives a map M, which contains binds
  of translated keys to functions to run - returns a function to
  receive events as such."
  [m]
  (fn [e]
    (let [key (e->key e)]
      ;; (log/debug "Key press event received: " e)
      ;; (log/debug "Translated e->key to: " key)
      (when (modkey? key)
        ;; (log/debug "Modkey is held down, setting it.")
        (swap! *modkeys assoc-in [key] true))
      (->> (dispatch m key @*modkeys)
           (map #(%))
           doall))))
